src/CityOnPlanet.cpp

```diff
@@ -202,123 +312,345 @@ CityOnPlanet::~CityOnPlanet()
-	const vector3d mz = m * vector3d(0, 0, 1);
+	// TODO: allow specifying city flavors based on various parameters (faction, world type, set from custom system def, etc.)
+	m_cityType = &s_cityFlavours[m_rand.Int32(s_cityFlavours.size())];
 
-	Random rand;
-	rand.seed(seed);
+	CalcCityRadius(m_body);
 
-	int population = planet->GetSystemBody()->GetPopulation();
-	int cityradius;
+	// TODO: run in a task thread
+	Generate(station);
 
-	if (planet->GetSystemBody()->HasAtmosphere()) {
-		population *= 1000;
-		cityradius = (population < 200) ? 200 : ((population > START_SEG_SIZE) ? START_SEG_SIZE : population);
-	} else {
-		population *= 100;
-		cityradius = (population < 250) ? 250 : ((population > START_SEG_SIZE_NO_ATMO) ? START_SEG_SIZE_NO_ATMO : population);
-	}
-
-	citybuildinglist_t *buildings = &s_buildingList;
-	vector3d cent = p;
-	const int cellsize_i = 80;
-	const double cellsize = double(cellsize_i);						// current widest building = 92
-	const double bodyradius = planet->GetSystemBody()->GetRadius(); // cache for bodyradius value
-
-	static const int gmid = (cityradius / cellsize_i);
-	static const int gsize = gmid * 2;
-
-	assert((START_SEG_SIZE / cellsize_i) < 100);
-	assert((START_SEG_SIZE_NO_ATMO / cellsize_i) < 100);
-	uint8_t cellgrid[200][200];
-	std::memset(cellgrid, 0, sizeof(cellgrid));
-
-	// calculate the size of the station model
-	const int x1 = floor(aabb.min.x / cellsize);
-	const int x2 = ceil(aabb.max.x / cellsize);
-	const int z1 = floor(aabb.min.z / cellsize);
-	const int z2 = ceil(aabb.max.z / cellsize);
-
-	// Clear the cells where the station is
-	for (int x = 0; x <= gsize; x++) {
-		for (int z = 0; z <= gsize; z++) {
-			const int zz = z - gmid;
-			const int xx = x - gmid;
-			if (zz > z1 && zz < z2 && xx > x1 && xx < x2)
-				cellgrid[x][z] = 1;
-		}
+	AddStaticGeomsToCollisionSpace();
+}
+
+void CityOnPlanet::Generate(SpaceStation *station)
+{
+	PROFILE_SCOPED()
+	Profiler::Clock _genTimer;
+	_genTimer.Start();
+
+	// Create the acceleration grid structure
+	// ==========================================
+
+	// Retrieve the station parameters
+	uint8_t stationSize[2] = {};
+	GetModelSize(station->GetAabb(), stationSize);
+
+	uint32_t stationSizeMax = std::max<uint32_t>(stationSize[0] / 2, stationSize[1] / 2);
+
+	// update the allowed radius to ensure the spaceport isn't larger than the entire city
+	m_cityRadius += stationSizeMax * CELLSIZE;
+
+	// For now, just assume the city is a circle around the spaceport in the center
+	const uint32_t cityExtents = (m_cityRadius / CELLSIZE);
+	m_citySize = cityExtents * 2;
+	m_gridPitch = std::ceil(m_citySize / 8.0);
+
+	Log::Verbose("Generating City for spacestation {}", station->GetSystemBody()->GetName());
+	Log::Verbose("\tpopulation: {0} size {1}x{1}c (radius {2})",
+		m_body->GetPopulation(), m_citySize, m_cityRadius);
+
+
+	// ensure divisible by 8 bytes for fast 64-bit lookup
+	if (m_gridPitch & 7) {
+		m_gridPitch = (m_gridPitch + 7) & ~7;
```

Comment:
It seems to me that 7 is not a `uint32_t` literal, and especially ~7 is worrisome.

Reply:
Maybe even  `auto rem = m_gridPitch % 8; if (rem != 0) { m_gridPitch += 8 - rem; }`

Reply:
This is the general form of a round-to-power-of-two expression in C/C++; using modulo introduces additional division operations when all that is needed is an addition and two binary and operations. I'll look at explicitly specifying the type of the literal, though.